openpgp/v2/read.go

```diff
@@ -740,43 +740,76 @@ func verifyDetachedSignatureReader(keyring KeyRing, signed, signature io.Reader,
-	var collectedErrors []error
-	now := config.Now()
+// - Signature is not expired (unless a zero time is passed to
+//   explicitly ignore expiration).
+func checkSignatureDetails(pk *packet.PublicKey, signature *packet.Signature, now time.Time, config *packet.Config) error {
+	if config.RejectHashAlgorithm(signature.Hash) {
+		return errors.SignatureError("insecure hash algorithm: " + signature.Hash.String())
+	}
 
-	if config.RejectMessageHashAlgorithm(signature.Hash) {
-		return errors.SignatureError("insecure message hash algorithm: " + signature.Hash.String())
+	if pk.CreationTime.Unix() > signature.CreationTime.Unix() {
+		return errors.ErrSignatureOlderThanKey
 	}
 
-	if verifiedKey.PublicKey.CreationTime.Unix() > signature.CreationTime.Unix() {
-		collectedErrors = append(collectedErrors, errors.ErrSignatureOlderThanKey)
+	for _, notation := range signature.Notations {
+		if notation.IsCritical && !config.KnownNotation(notation.Name) {
+			return errors.SignatureError("unknown critical notation: " + notation.Name)
+		}
 	}
 
-	sigsToCheck := []*packet.Signature{signature, verifiedKey.PrimarySelfSignature}
+	if !now.IsZero() && signature.SigExpired(now) {
+		return errors.ErrSignatureExpired
+	}
+	return nil
+}
 
+// checkMessageSignatureDetails verifies the metadata of the signature.
+// It checks the criteria of checkSignatureDetails for the message
+// signature and all relevant binding signatures.
+// In addition, the message signature hash algorithm is checked against
+// config.RejectMessageHashAlgorithms.
+func checkMessageSignatureDetails(verifiedKey *Key, signature *packet.Signature, config *packet.Config) error {
+	if config.RejectMessageHashAlgorithm(signature.Hash) {
+		return errors.SignatureError("insecure message hash algorithm: " + signature.Hash.String())
+	}
+
+	sigsToCheck := []*packet.Signature{signature, verifiedKey.PrimarySelfSignature}
 	if !verifiedKey.IsPrimary() {
 		sigsToCheck = append(sigsToCheck, verifiedKey.SelfSignature, verifiedKey.SelfSignature.EmbeddedSignature)
 	}
+	var errs []error
 	for _, sig := range sigsToCheck {
-		for _, notation := range sig.Notations {
-			if notation.IsCritical && !config.KnownNotation(notation.Name) {
-				return errors.SignatureError("unknown critical notation: " + notation.Name)
-			}
+		var pk *packet.PublicKey
+		if sig == verifiedKey.PrimarySelfSignature || sig == verifiedKey.SelfSignature {
+			pk = verifiedKey.Entity.PrimaryKey
+		} else {
+			pk = verifiedKey.PublicKey
+		}
+		var time time.Time
+		if sig == signature {
+			time = config.Now()
+		} else {
+			time = signature.CreationTime
+		}
+		err := checkSignatureDetails(pk, sig, time, config)
+		if err != nil {
+			errs = append(errs, err)
 		}
 	}
 
-	if signature.SigExpired(now) {
-		return errors.ErrSignatureExpired
+	// Return errors.ErrSignatureExpired last as gopenpgp might ignore it.
+	for _, err := range errs {
+		if err != errors.ErrSignatureExpired {
```

Comment:
```suggestion
		if errors.Is(err, errors.ErrSignatureExpired ) {
```
In the newer versions of go it is [recommended](https://pkg.go.dev/errors) to use `errors.Is` to compare errors. However, there might be a module naming clash here.

Reply:
In general you're right but in this particular case `errors.ErrSignatureExpired` is returned as-is in the function above, so there's no chance it'll be a wrapped error. So I think the existing code is safe and IMHO simpler in this case.

Reply:
Fine with me